### DISCUSSION TITLE
How to package an electron app / frontend / API together

### DISCUSSION THREAD

**Discussion by daiky00** [bennymeg/nx-electron#180] [Q&A] [answered]
I have the following folder structure
![Screen Shot 2022-10-05 at 12 42 00 PM](https://user-images.githubusercontent.com/15005162/194115323-48d37f57-70f2-4661-b55e-3214d0dc113f.png)

I was wondering how can I deploy this into an electron app installer all together can somebody please tell me what are the steps necessary to make this happen? 

Important note: The API works in local machine doesn't need to be online.

@bennymeg Your help will be greatly appreciated

**Comment by bennymeg** [comment, marked as answer]
1. In your _angular.json_ or _worksapce .json_ file, at your electron app build options, add an `additionalEntryPoints` entry to your api folder so it will be compiled alongside your electron & web apps.
2. In your electron app project, you can use `child_process.spawn` (or any other method you prefer) to spawn a new node process with your api, using a relative path to the main.js from your electron app (note that the relative paths between your apps in the app folder should be the same as your relative paths to the builds in the dist folder).
3. Build your project. 
4. In your package or maker options, add an `extraFiles` entry to account for copying over the directory with your compiled api into the final build.
5. Package your project.

**Reply by bennymeg** [reply]
Thats fine, you can define the `additionalEntryPoints` entry in your `workspace.json` instead (or compile you api manually).

**Reply by daiky00** [reply]
Thanks for the help

**Reply by daiky00** [reply]
@bennymeg the second step can you please tell me where do I need to initiate child_process.spawn and how?
and can you tell me if this `workspace.json(project.json)` looks correct to you for the electron app?
```JS
{
  "sourceRoot": "apps/sayf-identity-electron/src",
  "projectType": "application",
  "targets": {
    "build": {
      "executor": "nx-electron:build",
      "outputs": [
        "{options.outputPath}"
      ],
      "options": {
        "outputPath": "dist/apps/sayf-identity-electron",
        "main": "apps/sayf-identity-electron/src/main.ts",
        "tsConfig": "apps/sayf-identity-electron/tsconfig.app.json",
        "additionalEntryPoints": [
          {
            "entryName": "sayf-identity-api",
            "entryPath": "apps/sayf-identity-api/src/main.ts"
          }
        ],
        "assets": [
          "apps/sayf-identity-electron/src/assets"
        ]
      },
      "configurations": {
        "production": {
          "optimization": true,
          "extractLicenses": true,
          "inspect": false,
          "fileReplacements": [
            {
              "replace": "apps/sayf-identity-electron/src/environments/environment.ts",
              "with": "apps/sayf-identity-electron/src/environments/environment.prod.ts"
            }
          ]
        }
      }
    },
    "serve": {
      "executor": "nx-electron:execute",
      "options": {
        "buildTarget": "sayf-identity-electron:build"
      }
    },
    "package": {
      "executor": "nx-electron:package",
      "options": {
        "name": "sayf-identity-electron",
        "frontendProject": "sayf-identity-web",
        "extraFiles": [ "sayf-identity-api" ],
        "outputPath": "dist/packages",
        "prepackageOnly": true
      }
    },
    "make": {
      "executor": "nx-electron:make",
      "options": {
        "name": "sayf-identity-electron",
        "frontendProject": "sayf-identity-web",
        "extraFiles": [ "sayf-identity-api" ],
        "outputPath": "dist/executables"
      }
    },
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": [
        "{options.outputFile}"
      ],
      "options": {
        "lintFilePatterns": [
          "apps/sayf-identity-electron/**/*.ts"
        ]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": [
        "coverage/apps/sayf-identity-electron"
      ],
      "options": {
        "jestConfig": "apps/sayf-identity-electron/jest.config.js",
        "passWithNoTests": true
      }
    }
  }
}

```

**Reply by daiky00** [reply]
@bennymeg I will really appreciated your help with this part.

**Reply by daiky00** [reply]
@bennymeg This is urgent for me please help

**Reply by bennymeg** [reply]
Thats looks correct.
You can spawn the API child process at the `app.ts` module.

**Reply by daiky00** [reply]
@bennymeg but how?. Sorry I am not an expert on building electron apps :S

**Reply by daiky00** [reply]
@bennymeg I try doing the following and it doesn't work please help :(
```JS
  private static loadMainWindow() {
    // load the index.html of the app.
    if (!App.application.isPackaged) {
      App.mainWindow.loadURL(`http://localhost:${rendererAppPort}`);
    } else {
      const pathAPI =   format({
        pathname: join(__dirname, '..', apiAppName, 'main.js'),
        protocol: 'file:',
        slashes: true,
      })
      spawn('node', [pathAPI], { shell: true })
      App.mainWindow.loadURL(
        format({
          pathname: join(__dirname, '..', rendererAppName, 'index.html'),
          protocol: 'file:',
          slashes: true,
        })
      );
    }
  }
  ```

**Reply by daiky00** [reply]
if there anything you can see that is wrong from the code above let me know please!

**Reply by bennymeg** [reply]
It should be something like that (just outside the if statement), probably the path is incorrect.
1. Verify that the API is being built.
2. Write down the path and the file name of the API.
3. Spawn the API with the correct path.
4. Repeat steps 1-3 for the packaged application.

**Reply by daiky00** [reply]
@bennymeg the API is been built correctly but I put it in a way that it doesn't compile inside the electron app folder by changing `additionalEntryPoints` in the electron build options instead it builds in it is own folder image below. Now for the path are you saying I should put something like `../app/sayf-identify-api/main.js` ?
```JS
"additionalEntryPoints": [
          {
            "entryName": "../sayf-identity-api/main",
            "entryPath": "apps/sayf-identity-api/src/main.ts"
          }
        ],
  ```
<img width="316" alt="Screen Shot 2022-10-14 at 9 34 29 AM" src="https://user-images.githubusercontent.com/15005162/195859925-c04959e8-b1be-4e65-9a60-d4c71d8dce27.png">

**Reply by daiky00** [reply]
@bennymeg I change it to this and is not working.

**Note:** apiAppName = 'sayf-identity-api'

```JS
 private static loadMainWindow() {
    // load the index.html of the app.
    const pathAPI =   `../${apiAppName}/main.js`;
    spawn('node', [pathAPI], { shell: true });
    if (!App.application.isPackaged) {
      App.mainWindow.loadURL(`http://localhost:${rendererAppPort}`);
    } else {
      App.mainWindow.loadURL(
        format({
          pathname: join(__dirname, '..', rendererAppName, 'index.html'),
          protocol: 'file:',
          slashes: true,
        })
      );
    }
  }
  ```

**Reply by daiky00** [reply]
@bennymeg  any help would be much appreciated
